File: ios/UI/Sources/ShareSheet.swift
import SwiftUI

/// Who can see this list, and how to let somebody else in.
///
/// A link rather than an address: the server knows people by their Google account and
/// has no way to look one up by email, so an invitation is something you send by
/// whatever you already use to talk to them.
struct ShareSheet: View {
    let list: List
    let api: API
    /// Called after anything changes, so the screen behind can catch up.
    let changed: () async -> Void

    @Environment(\.dismiss) private var dismiss
    @State private var people: [Person] = []
    @State private var me: Int64?
    @State private var link: URL?
    @State private var loaded = false
    @State private var error: String?
    @State private var leaving = false

    private var iOwnIt: Bool { list.role >= .owner }

    var body: some View {
        NavigationStack {
            SwiftUI.List {
                if let link {
                    Section {
                        // Shown once. Only the hash is stored, so a link that is lost
                        // is remade rather than looked up.
                        Text(link.absoluteString)
                            .font(.footnote.monospaced())
                            .textSelection(.enabled)
                            .accessibilityIdentifier("share.link")

                        Button("Copy link") { copy(link.absoluteString) }
                            .accessibilityIdentifier("share.copy")
                    } header: {
                        Text("Send this to one person")
                    } footer: {
                        Text(
                            "It works once, for whoever opens it first, and expires "
                                + "in a week. It is shown only now."
                        )
                    }
                }

                Section("Who can see it") {
                    if !loaded {
                        ProgressView()
                    }
                    ForEach(people) { person in
                        HStack {
                            VStack(alignment: .leading, spacing: 1) {
                                Text(person.shown)
                                if person.name != nil, let email = person.email {
                                    Text(email)
                                        .font(.caption)
                                        .foregroundStyle(.secondary)
                                }
                            }
                            Spacer()
                            Text(person.userID == me ? "you" : person.role.rawValue)
                                .font(.caption)
                                .foregroundStyle(.secondary)
                        }
                        .accessibilityIdentifier("share.person.\(person.userID)")
                        .swipeActions(edge: .trailing) {
                            // An owner may remove anybody but themselves; there is no
                            // transfer, so a list without its owner is a list nobody
                            // could rename or delete.
                            if iOwnIt, person.role < .owner {
                                Button(role: .destructive) {
                                    Task { await act { try await api.remove(person, from: list) } }
                                } label: {
                                    Label("Remove", systemImage: "person.badge.minus")
                                }
                            }
                        }
                    }
                }

                if iOwnIt {
                    Section {
                        Button("Create a link") {
                            Task { await act { link = try await api.invite(to: list) } }
                        }
                        .accessibilityIdentifier("share.invite")

                        Button("Withdraw all links", role: .destructive) {
                            Task {
                                await act { try await api.revokeInvites(to: list) }
                                link = nil
                            }
                        }
                        .accessibilityIdentifier("share.revoke")
                    } footer: {
                        Text(
                            "Withdrawing cancels every link not yet used. People "
                                + "already on the list stay."
                        )
                    }
                } else {
                    Section {
                        Button("Leave this list", role: .destructive) { leaving = true }
                            .accessibilityIdentifier("share.leave")
                    } footer: {
                        Text("You will need a new link to come back.")
                    }
                }
            }
            .navigationTitle("Share \(list.name)")
            #if os(iOS)
                .navigationBarTitleDisplayMode(.inline)
            #endif
            .toolbar {
                ToolbarItem(placement: .confirmationAction) {
                    Button("Done") { dismiss() }
                        .accessibilityIdentifier("share.done")
                }
            }
            .confirmationDialog(
                "Leave \(list.name)?",
                isPresented: $leaving,
                titleVisibility: .visible
            ) {
                Button("Leave", role: .destructive) {
                    Task {
                        // Dismissed first: the sheet is about a list this person is
                        // about to stop being able to see.
                        dismiss()
                        await act { try await api.remove(mine(), from: list) }
                    }
                }
                Button("Cancel", role: .cancel) {}
            } message: {
                Text("It stays on the list for everyone else.")
            }
            .alert("Something went wrong", isPresented: .constant(error != nil)) {
                Button("OK") { error = nil }
            } message: {
                Text(error ?? "")
            }
            .task { await load() }
        }
        .frame(minWidth: 380, minHeight: 420)
    }

    /// Me, as a `Person`, so leaving is removing like any other.
    private func mine() -> Person {
        people.first { $0.userID == me }
            ?? Person(userID: me ?? 0, name: nil, email: nil, role: list.role)
    }

    private func copy(_ text: String) {
        #if os(macOS)
            NSPasteboard.general.clearContents()
            NSPasteboard.general.setString(text, forType: .string)
        #else
            UIPasteboard.general.string = text
        #endif
    }

    /// Runs something, reports what went wrong, and lets the screen behind catch up.
    private func act(_ work: () async throws -> Void) async {
        do {
            try await work()
            await load()
            await changed()
        } catch let problem as APIError {
            error = problem.localizedDescription
        } catch {
            self.error = error.localizedDescription
        }
    }

    private func load() async {
        do {
            async let people = api.people(on: list)
            async let me = api.whoAmI()
            (self.people, self.me) = try await (people, me.id)
            error = nil
        } catch let problem as APIError {
            error = problem.localizedDescription
        } catch {
            self.error = error.localizedDescription
        }
        loaded = true
    }
}

File: android/app/src/main/kotlin/com/cernauskas/shoppinglist/ui/ShareSheet.kt
package com.cernauskas.shoppinglist.ui

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ContentCopy
import androidx.compose.material.icons.filled.PersonRemove
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import androidx.compose.ui.unit.dp
import com.cernauskas.shoppinglist.data.Person
import com.cernauskas.shoppinglist.data.Role
import com.cernauskas.shoppinglist.data.ShoppingList

/**
 * Who can see this list, and how to let somebody else in.
 *
 * A link rather than an address: the server knows people by their Google account and
 * cannot look one up by email, so an invitation is something you send by whatever you
 * already use to talk to them.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ShareSheet(list: ShoppingList, model: ListsViewModel, onDismiss: () -> Unit) {
    val context = LocalContext.current
    // The composable's own scope, not the view model's: `viewModelScope` is the view
    // model's business, and this work belongs to the sheet that is on screen.
    val scope = rememberCoroutineScope()
    var people by remember { mutableStateOf<List<Person>>(emptyList()) }
    var me by remember { mutableStateOf<Long?>(null) }
    var link by remember { mutableStateOf<String?>(null) }
    var loading by remember { mutableStateOf(true) }
    var leaving by remember { mutableStateOf(false) }

    val iOwnIt = list.role >= Role.OWNER

    suspend fun refresh() {
        people = model.people(list)
        me = model.whoAmI()
        loading = false
    }

    LaunchedEffect(list.id) {
        try {
            refresh()
        } catch (e: Exception) {
            loading = false
            model.say(e.message ?: "Could not read who this is shared with.")
        }
    }

    // Read again whenever the list changes, for as long as the sheet is up.
    //
    // Somebody following the link is the one change this sheet exists to show, and it
    // arrives while the sheet is open -- you hand over the code and watch. Loading the
    // members once meant the sheet still said "who can see it: you" after they were
    // already reading the list, which reads as the code not having worked.
    LaunchedEffect(list.id) {
        while (true) {
            try {
                model.watchList(list).collect {
                    try {
                        refresh()
                    } catch (_: Exception) {
                        // A failed re-read leaves the names that are already up, which
                        // is a moment out of date rather than blank.
                    }
                }
            } catch (_: Exception) {
                // Losing the stream is ordinary. Nothing to say about it here.
            }
            delay(3_000)
        }
    }

    // Fully expanded and scrollable. Once a link is showing, the members and the
    // buttons under it sit past the bottom of a half-height sheet, and Withdraw is a
    // control nobody can reach.
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true),
    ) {
        Column(
            Modifier
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 24.dp)
                .padding(bottom = 24.dp)
                .navigationBarsPadding(),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            Text("Share ${list.name}", style = MaterialTheme.typography.titleLarge)

            link?.let { code ->
                Card(Modifier.fillMaxWidth()) {
                    Column(Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(8.dp)) {
                        Text(
                            code,
                            style = MaterialTheme.typography.bodySmall,
                            fontFamily = androidx.compose.ui.text.font.FontFamily.Monospace,
                        )
                        Text(
                            "Works once, for whoever uses it first, and expires in a "
                                + "week. Shown only now.",
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                        FilledTonalButton(
                            onClick = { context.copy(code); model.say("Code copied") },
                            modifier = Modifier.fillMaxWidth(),
                        ) {
                            Icon(Icons.Default.ContentCopy, contentDescription = null)
                            Spacer(Modifier.width(8.dp))
                            Text("Copy code")
                        }
                    }
                }
            }

            Text("Who can see it", style = MaterialTheme.typography.titleSmall)

            if (loading) {
                LinearProgressIndicator(Modifier.fillMaxWidth())
            }

            people.forEach { person ->
                ListItem(
                    headlineContent = { Text(person.shown) },
                    supportingContent = {
                        Text(if (person.userId == me) "you" else person.role.name.lowercase())
                    },
                    trailingContent = {
                        // An owner may remove anybody but themselves: there is no
                        // transfer, so a list without its owner is one nobody could
                        // rename or delete.
                        if (iOwnIt && person.role < Role.OWNER) {
                            IconButton(onClick = {
                                scope.attempt(model) {
                                    model.remove(person, list)
                                    refresh()
                                }
                            }) {
                                Icon(
                                    Icons.Default.PersonRemove,
                                    contentDescription = "Remove ${person.shown}",
                                )
                            }
                        }
                    },
                )
            }

            if (iOwnIt) {
                Button(
                    onClick = {
                        scope.attempt(model) { link = model.invite(list) }
                    },
                    modifier = Modifier.fillMaxWidth(),
                ) {
                    Text("Create a code")
                }
                TextButton(
                    onClick = {
                        scope.attempt(model) {
                            model.revokeInvites(list)
                            link = null
                            model.say("Every unused link withdrawn")
                        }
                    },
                    modifier = Modifier.fillMaxWidth(),
                ) {
                    Text("Withdraw all codes")
                }
            } else {
                TextButton(onClick = { leaving = true }, modifier = Modifier.fillMaxWidth()) {
                    Text("Leave this list")
                }
            }
        }
    }

    if (leaving) {
        AlertDialog(
            onDismissRequest = { leaving = false },
            title = { Text("Leave ${list.name}?") },
            text = { Text("It stays on the list for everyone else. You will need a new link to come back.") },
            confirmButton = {
                TextButton(onClick = {
                    leaving = false
                    onDismiss()
                    val mine = people.firstOrNull { it.userId == me }
                    if (mine != null) {
                        scope.attempt(model) {
                            model.remove(mine, list)
                            model.load()
                        }
                    }
                }) { Text("Leave") }
            },
            dismissButton = { TextButton(onClick = { leaving = false }) { Text("Cancel") } },
        )
    }
}

private fun Context.copy(text: String) {
    val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
    clipboard.setPrimaryClip(ClipData.newPlainText("Shopping list link", text))
}

/** Runs something that may fail, and says so rather than crashing the screen. */
private fun CoroutineScope.attempt(model: ListsViewModel, work: suspend () -> kotlin.Unit) {
    launch {
        try {
            work()
        } catch (e: Exception) {
            model.say(e.message ?: "Something went wrong.")
        }
    }
}
